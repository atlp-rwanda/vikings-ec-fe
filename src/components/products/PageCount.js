import React from 'react';
import next from '../../../public/images/next.png';
import prevPage from '../../../public/images/nextPage.svg';
import nextPage from '../../../public/images/prevPage.svg';

const PageCount = ({ ...props }) => (
  <div className={props.className} data-testid="page-number">
    <div className="flex xs:gap-[5px] md:gap-[140px]  justify-center items-center">
      <div>
        <button
          data-testid="next-page"
          type="button"
          className="flex justify-center items-center gap-2 md:px-4 md:py-2 lg:px-4 lf:py-2 xl:px-4 xl:py-2 xs:px-1 xs:py-1 hover:bg-[#045504] bg-[#10c510] text-[white] xs:w-[150px]"
          onClick={() => props.click(true)}
        >
          Next page
          <img src={next} alt="next page" className="mt-2" />
        </button>
      </div>
      <div className="flex gap-[30px]">
        <p>Page</p>
        <span className="ml-[-20px]">{props.currentPage}</span>
        <p className="ml-[-20px]">
          of
          <span className="ml-[13px]">{props.totalPages}</span>
        </p>
        <button type="button" className="flex justify-center items-center gap-[20px] hover:bg-[#045504] bg-[#10c510] h-[23px] px-2 py-2">
          <img src={prevPage} onClick={() => props.click(false)} alt="previous page" />
          <img src={nextPage} onClick={() => props.click(true)} alt="next-page" />
        </button>
      </div>
    </div>
  </div>
);

export default PageCount;
