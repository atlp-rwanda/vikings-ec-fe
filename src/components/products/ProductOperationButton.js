import React from 'react';

const ProductOperationButton = ({ ...props }) => (
  <button
    data-testid={props.testId}
    onClick={props.onClick}
    className={props.className}
    type="button"
    title={props.title}
    onClick={props.onClick}
    onMouseEnter={props.handleMouseEnter}
    onMouseLeave={props.handleMouseLeave}
  >
    <img
      src={props.icon}
      alt={props.alt}
      className={props.size}
    />
  </button>
);
export default ProductOperationButton;
