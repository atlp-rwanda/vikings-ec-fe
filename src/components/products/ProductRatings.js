import React from 'react';
import star from '../../../public/images/star.svg';
import star1 from '../../../public/images/star1.svg';

const ProductRatings = ({ rate }) => {
  const ratingsRange = {
    one: 0, two: 0, three: 0, four: 0, five: 0,
  };
  for (let i = 0; i < rate.length; i++) {
    if (rate[i].rate === 1) {
      ratingsRange.one += 1;
    } else if (rate[i].rate === 2) {
      ratingsRange.two += 1;
    } else if (rate[i].rate === 3) {
      ratingsRange.three += 1;
    } else if (rate[i].rate === 4) {
      ratingsRange.four += 1;
    } else if (rate[i].rate === 5) {
      ratingsRange.five += 1;
    }
  }
  let totalUsers = 0;
  let totalRatings = 0;
  let count = 1;
  for (const key in ratingsRange) {
    totalUsers += ratingsRange[key];
    totalRatings += ratingsRange[key] * count;
    count++;
  }
  const averageRatings = Math.floor(totalRatings / totalUsers) || 0;
  return (
    <div className="flex">
      {[...Array(averageRatings)].map((_, index) => (
        <img key={index} src={star} alt="star" className="w-5 h-5" />
      ))}
      {[...Array(5 - averageRatings)].map((_, index) => (
        <img key={index} src={star1} alt="star" className="w-5 h-5" />
      ))}
    </div>
  );
};
export default ProductRatings;
