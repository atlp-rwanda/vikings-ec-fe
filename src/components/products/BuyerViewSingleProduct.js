import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ProductRatings from './ProductRatings';
import Reviewers from './Reviewers';
import loader from '../../../public/images/icons/loader.svg';
import { getProductRatings } from '../../features/product/getRatingsSlice';

const BuyerViewProduct = ({ ...props }) => {
  const dispatch = useDispatch();
  const { ratings, isLoading } = useSelector((state) => state.getRatings);
  const [reviews, setReviews] = useState([]);

  useEffect(() => {
    if (props.products && props.products.id) {
      dispatch(getProductRatings(props.products.id));
    }
  }, [props.products && props.products.id]);

  useEffect(() => {
    if (Array.isArray(ratings) && ratings.length > 0) {
      const reviewers = ratings.map((rate) => ({
        buyer: rate.buyer,
        feedback: rate.feedback,
      }));
      setReviews(reviewers);
    }
  }, [ratings]);
  let reviewers = null;
  if (Array.isArray(ratings) && ratings.length > 0) {
    reviewers = reviews.map((r, index) => (
      <Reviewers key={index} reviewer={r} />
    ));
  }
  const fullIViewmage = props.products.images && (
    <img
      src={props.products.images[props.fullImage]}
      alt="product"
      className="xs:w-full lg:w-full sm:w-full xs:ml-[-16px] md:w-full md:h-[250px] object-cover"
    />
  );
  const otherImages =
    props.products.images &&
    props.products.images.map(
      (image, i) =>
        i !== props.fullImage && (
          <img src={image} key={i} alt="product" className="w-24 h-20 mt-2" />
        )
    );
  return (
    <div>
      <div className="md:flex lg:flex flex-wrap md:flex-nowrap w-[100%] mt-24 md:gap-10 lg:gap-4">
        <div className="md:w-[900px] xs:w-[108%] lg:w-1/2 sm:ml-[-13px] xs:ml-[0px] px-4 ">
          {fullIViewmage}
          <div className="md:mt-[-190px]">{props.switchImages}</div>
        </div>
        <div className="w-full md:w-1/1">
          <p className="text-black font-bold text-[20px] pb-2">
            {props.products.name}
          </p>
          <p className="text-[#464c55] text-[18px] font-bold pb-2">
            ${props.products.price}
          </p>
          <div className="space-y-[-10px]">
            {ratings ? (
              <ProductRatings rate={ratings} />
            ) : (
              <img
                src={loader}
                alt="Loader Spinner"
                className="text-green-500 animate-spin w-[25px] text-center"
                data-testid="spinner"
              />
            )}
            <div className="flex space-x-[25px] pb-[40px]">
              {props.wish}
              {props.addCart}
            </div>
            <div className="lg:flex lg:gap-3 md:flex md:gap-3 pb-2 grid sm:grid-cols-3 xs:grid-cols-3 sm:space-y-2 xs:space-y-2">
              {otherImages}
            </div>
          </div>
        </div>
      </div>
      <div className="w-[100%] md:space-y-3">
        <p className="text-[#393d3e] pt-8 font-bold text-[18px] italic">
          Reviews
        </p>
        <div className="md:space-y-3">{reviewers || 'No reviews found'}</div>
      </div>
    </div>
  );
};

export default BuyerViewProduct;