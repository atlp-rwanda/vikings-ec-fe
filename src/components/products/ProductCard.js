import React, { useEffect, useState } from 'react';
import ProductRatings from './ProductRatings';
import getDiscount from '../../utils/getDiscount';

const ProductCard = ({ getPersistCartSetter, ...props }) => {
  const [hovered, setHovered] = useState(false);
  const [persistCart, setPersistCart] = useState(false);
  useEffect(() => {
    if (getPersistCartSetter) {
      getPersistCartSetter((value) => {
        setPersistCart(value);
        if (value) setHovered(false);
      });
    }
  }, [getPersistCartSetter]);
  const singleProduct = () => {
    if (props.wish) {
      window.location.href = `/products/${props.product.id}`;
    } else {
      props.viewSingleProduct(
        props.product.name,
        props.product.images[0],
        props.product.price,
        props.product.quantity,
        props.product.id,
        props.product.bonus,
        props.product.isAvailable,
      );
    }
  };

  return (
    <div
      data-testid="product-card"
      className="flex-col border border-gray-100 shadow-sm relative"
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}

    >
      {
        (props.product.bonus > 0) && (
          <div className="bg-red-500 text-white absolute top-6 left-6 w-fit p-2 text-[12px]">
            -
            {getDiscount(props.product.price, props.product.bonus)}
            {' '}
            %
          </div>
        )
      }
      <div className="h-[200px] md:h-[300px] overflow-y-clip">
        <img
          src={props.product.images[0]}
          alt="product"
          className="w-full object-cover cursor-pointer h-full"
          onClick={() => singleProduct()}
        />
        {(hovered || persistCart) && (
          <div className="absolute bottom-16 right-4">
            {props.wish}
            {props.addCart}
          </div>
        )}
        {hovered && (
        <div className="absolute bottom-16 right-4 flex gap-2 md:bottom-24">
          {props.editBTN}
          {props.deleteBTN}
        </div>
        )}
      </div>
      <div className="flex justify-between items-center bg-white p-2">
        <p
          data-testid="product-name"
          className="text-indigo-900 text-[20px] mt-3 cursor-pointer hover:text-[#099f09] truncate"
          onClick={() => singleProduct()}
        >
          {props.product.name}
        </p>
        <div className="flex justify-between mt-2 items-center">
          <div className="text-[18px] text-[#099f09] font-bold truncate">
            {props.product.price}
            {' '}
            RWF
          </div>
        </div>
      </div>
    </div>
  );
};

export default ProductCard;
