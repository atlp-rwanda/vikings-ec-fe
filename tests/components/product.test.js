import React from 'react';
import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { expect, describe, it } from '@jest/globals';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import axiosMock from 'axios';
import BuyerViewSingleProduct from '../../src/components/products/BuyerViewSingleProduct';
import PageCount from '../../src/components/products/PageCount';
import ProductOperationButton from '../../src/components/products/ProductOperationButton';
import SwitchImages from '../../src/components/products/SwitchImages';
import productCardMock, {
  mockProductsWithWishBtn,
} from '../mocks/products/product.mock';
import ProductCard from '../../src/components/products/ProductCard';
import SellerViewSingleProduct from '../../src/components/products/SellerViewSingleProduct';
import Products from '../../src/pages/dashboard/Products';
import SingProductPage from '../../src/pages/SingleProductPage';
import HomePage from '../../src/pages/HomePage';
import { getProductList } from '../../src/features/product/getProductsSilice';
import getMessage from '../../src/features/actions/welcomeAction';
import Reviewers from '../../src/components/products/Reviewers';
import ProductRatings from '../../src/components/products/ProductRatings';
import st from '../../src/store';

describe('ProductCard', () => {
  it('should render recommended products', () => {
    const { getByText } = render(
      <Provider store={st}>
        <MemoryRouter>
          <SingProductPage />
        </MemoryRouter>
      </Provider>
    );
    expect(getByText('Recommended For You')).toBeInTheDocument();
  });
  it('should render product card with correct props', () => {
    render(<ProductCard {...productCardMock} />);
    expect(screen.getByAltText('product')).toBeInTheDocument();
    expect(screen.getByText('product name')).toBeInTheDocument();
    expect(screen.getByText('100 RWF')).toBeInTheDocument();
  });

  it('should call viewSingleProduct when clicking on product name', () => {
    render(<ProductCard {...productCardMock} />);
    const clickName = screen.getByTestId('product-name');
    clickName.click();
    expect(productCardMock.viewSingleProduct).toHaveBeenCalledWith(
      'product name',
      'product-image.jpg',
      100,
      1,
      '12345',
      2000,
      true
    );
  });

  it('should call window location for single product', () => {
    render(<ProductCard {...mockProductsWithWishBtn} />);
    const productName = screen.getByText('product name');
    const clickName = screen.getByTestId('product-name');
    clickName.click();
    expect(productName).toBeInTheDocument();
  });

  it('should set hovered state', () => {
    const { getByTestId } = render(<ProductCard {...productCardMock} />);
    const cardElement = getByTestId('product-card');
    fireEvent.mouseEnter(cardElement);
    fireEvent.mouseLeave(cardElement);
    const productName = screen.getByText('product name');
    expect(productName).toBeInTheDocument();
  });
});

describe('DashboardPage', () => {
  const mockStore = configureStore([thunk]);
  it('should render loading spinner when isLoading is true', () => {
    const isLoading = true;
    const productsList = [];
    const initialState = { product: { productsList, isLoading } };
    const store = mockStore(initialState);
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Products isLoading={isLoading} products={[]} />
        </BrowserRouter>
      </Provider>,
    );
  });
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn().mockReturnValue({
    search: '?id=1',
  }),
}));

describe('SingProductPage', () => {
  const mockStore = configureStore({})({
    singleProduct: {
      product: {
        name: 'Test Product',
        price: 100,
        images: ['image1.jpg', 'image2.jpg'],
      },
      isLoading: false,
    },
    getRatings: {
      ratings: null,
      isLoading: false,
    },
    provideRatings: {},
    recommendedProducts: {
      products: null,
      isLoading: false,
    },
    product: {
      productList: {},
      isLoading: false,
    },
  });
  it('should render SingProductPage component correctly', () => {
    const tree = renderer
      .create(
        <Provider store={mockStore}>
          <MemoryRouter>
            <SingProductPage />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('BuyerViewSingleProduct', () => {
  const product = {
    images: ['image1.jpg', 'image2.jpg', 'image3.jpg'],
    fullImage: 0,
    name: 'Product Name',
    price: 100,
  };

  it('renders product details correctly', () => {
    const mockStore = configureStore({})({
      singleProduct: {
        product: {
          name: 'Test Product',
          price: 100,
          images: ['image1.jpg', 'image2.jpg'],
        },
        isLoading: false,
      },
      getRatings: {
        ratings: null,
        isLoading: false,
      },
      provideRatings: {
        isLoading: false,
      },
    });
    const { getByText } = render(
      <Provider store={mockStore}>
        <BuyerViewSingleProduct products={product} />
      </Provider>,
    );
    expect(getByText('Product Name')).toBeInTheDocument();
    expect(getByText('100 RWF')).toBeInTheDocument();
  });
});

describe('PageCount', () => {
  const defaultProps = {
    className: 'test-class',
    currentPage: 1,
    totalPages: 10,
    click: jest.fn(),
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('shoud render PageCount correctly', () => {
    const { getByAltText } = render(<PageCount {...defaultProps} />);
    expect(getByAltText('previous page')).toBeInTheDocument();
  });

  it('calls click function with true when next page button is clicked', () => {
    const { getByText } = render(<PageCount {...defaultProps} />);
    const nextButton = getByText('Next page');
    fireEvent.click(nextButton);
    expect(defaultProps.click).toHaveBeenCalledWith(true);
  });

  it('calls click function with false when previous page button is clicked', () => {
    const { getByAltText } = render(<PageCount {...defaultProps} />);

    const prevButton = getByAltText('previous page');
    fireEvent.click(prevButton);
    expect(defaultProps.click).toHaveBeenCalledWith(false);
  });
  it('calls click function with false when next arrow button is clicked', () => {
    const { getByAltText } = render(<PageCount {...defaultProps} />);

    const nextButton = getByAltText('next-page');
    fireEvent.click(nextButton);
    expect(defaultProps.click).toHaveBeenCalledWith(true);
    expect(getByAltText('next-page')).toBeInTheDocument();
  });
});

describe('ProductOperationButton', () => {
  const defaultProps = {
    className: 'test-class',
    title: 'Test Title',
    handleMouseEnter: jest.fn(),
    handleMouseLeave: jest.fn(),
    icon: 'test-icon.png',
    alt: 'Test Alt',
    size: 'test-size',
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render roductOperationButton  correctly', () => {
    const { getByRole, getByAltText } = render(
      <ProductOperationButton {...defaultProps} />,
    );

    const buttonElement = getByRole('button');
    const iconElement = getByAltText('Test Alt');

    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveClass('test-class');
    expect(buttonElement).toHaveAttribute('type', 'button');
  });
});

describe('SellerViewSingleProduct', () => {
  const product = {
    image: 'product-image.jpg',
    name: 'Example Product',
    price: 100,
    quantity: 10,
  };
});

describe('SwitchImages', () => {
  const props = {
    className: 'switch-images-container',
    switchCurrentImage: jest.fn(),
  };

  it('should render left and right buttons with correct props', () => {
    const { getByTestId } = render(<SwitchImages {...props} />);

    const leftButton = getByTestId('left-button');
    const rightButton = getByTestId('right-button');

    expect(leftButton).toBeInTheDocument();
    expect(rightButton).toBeInTheDocument();

    fireEvent.click(leftButton);
    expect(props.switchCurrentImage).toHaveBeenCalledWith(false);

    fireEvent.click(rightButton);
    expect(props.switchCurrentImage).toHaveBeenCalledWith(true);
  });
});

const mockStore = configureStore([thunk]);

describe('HomePage Component', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      product: {
        productsList: {
          currentPage: 1,
          totalPages: 2,
          rows: [
            {
              id: 1,
              name: 'Product 1',
              images: ['image1.jpg', 'image2.jpg'],
              price: 100,
              quantity: 3,
            },
          ],
        },
        isLoading: false,
      },
      message: {
        data: { message: 'Welcome to the homepage' },
      },
      googleAuth: {
        data: { user: { id: 1, name: 'User' } },
      },
      getMessages: { isLoading: false, messages: [] },
      sendMessage: { isLoading: false },
    });
  });

  it('should render product cards when products are available', () => {
    store.dispatch(getMessage());
    store.dispatch(getProductList({ pageNumber: 1 }));
    const { getByText } = render(
      <Provider store={store}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </Provider>,
    );
    const productCard = screen.getByText('Product 1');
    fireEvent.mouseEnter(productCard);
    expect(productCard).toHaveClass(
      'text-indigo-900 text-[20px] mt-3 cursor-pointer hover:text-[#099f09]',
    );
    expect(getByText('Product 1')).toBeInTheDocument();
  });

  it('should render HomePage', () => {
    store.getState().product.isLoading = true;
    const { getAllByTestId } = render(
      <Provider store={store}>
        <MemoryRouter>
          <HomePage />
        </MemoryRouter>
      </Provider>,
    );
  });
});

const mockStore2 = configureStore([thunk]);
const store = mockStore2({
  markProduct: {
    data: null,
    isLoading: false
  },
  product: {
    productsList: {
      currentPage: 1,
      totalPages: 2,
      rows: [
        {
          id: 1,
          name: 'Product 1',
          images: ['image1.jpg', 'image2.jpg'],
          price: 10,
          quantity: 5,
          productId: 'product1',
        },
        {
          id: 2,
          name: 'Product 2',
          images: ['image1.jpg', 'images2.jpg'],
          price: 20,
          quantity: 10,
          productId: 'product2',
        },
      ],
    },
    isLoading: false,
  },
  getRatings: {
    ratings: null,
    isLoading: false,
  },
});

describe('DashboardPage', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Products />
        </BrowserRouter>
      </Provider>,
    );
  });

  it('should render the product cards on dashboard', () => {
    const product1Name = screen.getByText('Product 1');
    const product2Name = screen.getByText('Product 2');

    expect(product1Name).toBeInTheDocument();
    expect(product2Name).toBeInTheDocument();
  });

  it('sets selected product and product clicked state', () => {
    const { getAllByText } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Products />
        </BrowserRouter>
      </Provider>,
    );
    const product1Names = getAllByText('Product 1');
    fireEvent.click(product1Names[0]);
  });

  it('should change page number and dispatches getProductList with correct page number', () => {
    const dispatchMock = jest.fn();
    const getProductListMock = jest.fn();
    const productsList = {
      currentPage: 1,
      totalPages: 5,
    };
    const { getAllByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
          <Products dispatch={dispatchMock} getProductList={getProductListMock} productsList={productsList} />
        </BrowserRouter>
      </Provider>,
    );
    const testId = getAllByTestId('next-page');
    fireEvent.click(testId[0]);
    expect(productsList.currentPage).toBe(1);
  });
});

describe('Reviewers component', () => {
  const reviewer = {
    feedback: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    buyer: {
      firstname: 'John',
      lastname: 'Doe',
      avatar: 'http://example.com/avatar.png',
    },
  };

  it('renders the reviewer name', () => {
    const { getByText } = render(<Reviewers reviewer={reviewer} />);
    expect(
      getByText(`${reviewer.buyer.firstname} ${reviewer.buyer.lastname}`)
    ).toBeInTheDocument();
  });

  it('renders the reviewer feedback', () => {
    const { getByText } = render(<Reviewers reviewer={reviewer} />);
    expect(getByText(reviewer.feedback)).toBeInTheDocument();
  });

  it('renders the reviewer avatar when provided', () => {
    const { getByAltText } = render(<Reviewers reviewer={reviewer} />);
    expect(getByAltText('profile')).toHaveAttribute(
      'src',
      reviewer.buyer.avatar
    );
  });
});

describe('ProductRatings component', () => {
  const ratings = [
    { ratings: 1 },
    { ratings: 3 },
    { ratings: 5 },
    { ratings: 4 },
    { ratings: 5 },
    { ratings: 2 },
    { ratings: 4 },
  ];

  it('renders the correct number of stars for the average rating', () => {
    const { container } = render(<ProductRatings rate={ratings} />);
    const stars = container.querySelectorAll('img');
    expect(stars).toHaveLength(5);
  });
});
