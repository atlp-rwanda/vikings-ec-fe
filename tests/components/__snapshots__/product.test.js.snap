// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingProductPage should render SingProductPage component correctly 1`] = `
<div
  data-test-id="recommended"
>
  <div
    className="px-24 xs:px-2"
  >
    <div>
      <div
        className="md:flex gap-24"
      >
        <div
          className="relative h-[250px] overflow-y-clip md:w-[500px] md:h-[350px]"
        >
          <img
            alt="product"
            className="md:w-full object-center"
            src="image1.jpg"
          />
          <div
            className="absolute top-[44%] w-full"
          >
            <div
              className="flex justify-between"
            >
              <button
                className="bg-white hover:bg-[#0cc70c] w-[40px] h-[40] flex justify-center items-center rounded-full"
                data-testid="right-button"
                onClick={[Function]}
                type="button"
              >
                <img
                  alt="left"
                  src=""
                />
              </button>
              <button
                className="bg-white  hover:bg-[#0cc70c] w-[40px] h-[40px] flex justify-center items-center rounded-full"
                data-testid="left-button"
                onClick={[Function]}
                type="button"
              >
                <img
                  alt="right"
                  src=""
                />
              </button>
            </div>
          </div>
        </div>
        <div
          className=""
        >
          <p
            className="text-black font-bold text-[20px] pb-2"
          >
            Test Product
          </p>
          <p
            className="text-[#464c55] text-[18px] font-bold pb-2"
          >
            100
             
            RWF
          </p>
          <div
            className=""
          >
            <img
              alt="Loader Spinner"
              className="text-green-500 animate-spin w-[25px] text-center"
              data-testid="spinner"
              src=""
            />
            <div
              className="flex gap-4"
            >
              <button
                className="mt-[20px] hover:bg-[#22f122] bg-[#f6f4f4] h-[32px] w-[32px] rounded-full flex justify-center items-center"
                title="Wish product"
                type="button"
              >
                <img
                  alt="wish"
                  src=""
                />
              </button>
              <div
                className="ease-linear transition-all duration-150"
                onClick={[Function]}
              >
                <button
                  className="mt-[20px] bg-[#f6f4f4] hover:bg-[#099f09] h-[32px] w-[32px] rounded-full flex justify-center items-center hover:bg-[#D9F2D9]"
                  disabled={true}
                  title="Product is not available"
                  type="button"
                >
                  <img
                    alt="shop"
                    src=""
                  />
                </button>
              </div>
            </div>
            <div
              className="flex justify-between h-fit overflow-clip items-center mt-7 gap-3"
            >
              <img
                alt="product"
                className="w-24 h-20 mt-2 object-cover "
                src="image2.jpg"
              />
            </div>
            <div
              className="w-fit mt-6"
            >
              <button
                className=" w-full my-2 h-[45px] rounded-md bg-[#338E03] py-2 px-3 text-sm font-semibold text-white hover:bg-[#64B937] bg-green-500 hover:bg-green-600 rounded-none"
                onClick={[Function]}
              >
                Add review
                 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="w-[100%] md:space-y-3"
      >
        <p
          className="text-[#393d3e] pt-8 font-bold text-[18px] italic"
        >
          Reviews
        </p>
        <div
          className="md:space-y-3"
        >
          No reviews found
        </div>
        <form
          className="fixed top-0 left-0 flex flex-col justify-center items-center px-4 w-full h-full bg-black bg-opacity-50 z-50 hidden"
          onSubmit={[Function]}
        >
          <div
            className="max-w-2xl my-auto backdrop-blur-sm z-50 p-4 border border-grey-400 bg-slate-200 poppins w-full px-4 bg-opacity-80 "
            id="container"
          >
            <h2>
              Provide Ratings and Feedback
            </h2>
            <div
              className="flex items-center"
            >
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  {
                    "color": "orange",
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  {
                    "color": "orange",
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  {
                    "color": "orange",
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  {
                    "color": "orange",
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
                />
              </svg>
              <svg
                fill="currentColor"
                height="1em"
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  {
                    "color": "orange",
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 0 0 .6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0 0 46.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3zM664.8 561.6l36.1 210.3L512 672.7 323.1 772l36.1-210.3-152.8-149L417.6 382 512 190.7 606.4 382l211.2 30.7-152.8 148.9z"
                />
              </svg>
            </div>
            <div
              className="mt-5"
            >
              <div
                className="h-[80px] undefined"
              >
                <label
                  className="undefined sr-only""
                  htmlFor="Feedback"
                >
                  Feedback
                </label>
                <div
                  className="w-full"
                >
                  <textarea
                    className="block p-2.5 w-full text-sm text-gray-900 bg-gray-50 rounded-lg border border-gray-300 focus:border-blue-50"
                    id="feedback"
                    label="Feedback"
                    name="Feedback"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Provide your feedback"
                    rows="4"
                  />
                </div>
              </div>
            </div>
            <div
              className="py-3 sm:px-6 flex gap-2 flex-row-reverse"
            >
              <div
                className="flex xs:block"
              >
                <button
                  className=" w-full my-2 h-[45px] rounded-md bg-[#338E03] py-2 px-3 text-sm font-semibold text-white hover:bg-[#64B937] text-white sm:ml-3 bg-green-500 hover:bg-green-600 px-5 py-2.5 mt-20 font-bold rounded-none"
                  onClick={[Function]}
                  role="submit"
                  type="submit"
                >
                  Add review
                   
                </button>
              </div>
              <div
                className="flex xs:block"
              >
                <button
                  className=" w-full my-2 h-[45px] rounded-md bg-[#338E03] py-2 px-3 text-sm font-semibold text-white hover:bg-[#64B937] text-white px-5 py-2.5 mt-20 font-bold rounded-none"
                  onClick={[Function]}
                  role="cancel"
                  type="button"
                >
                  cancel
                   
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  <div
    className="pb-4 mt-20"
  >
    <h1
      className="font-bold text-xl text-center"
    >
      Recommended For You
    </h1>
    <div
      className=" flex justify-center flex-wrap gap-8 mt-8 "
    />
  </div>
</div>
`;
